tests/unit/controller_test.cpp

```diff
@@ -0,0 +1,188 @@
+TEST_F(ControllerTest, NominalCheck1) {
+    // Arrange
+    // Set up any necessary variables, mocks, etc
+    w_status_t expected_status = W_SUCCESS;
+    float expected_angle = -0.1745;
+    flight_phase_get_state_fake.return_val = STATE_ACT_ALLOWED;
+
+    float p_dyn = 1e4;
+    float coeff = 1.0f;
+    controller_gain_t controller_gain = {0};
+    float roll_state_arr[FEEDBACK_GAIN_NUM] = {1, 1, 1};
+
+    // Act
+    // Call the function to be tested
+
+    w_status_t actual_status = interpolate_gain(p_dyn, coeff, &controller_gain);
+
+    float actual_angle;
+    get_commanded_angle(controller_gain, roll_state_arr, &actual_angle);
+
+    // Assert
+    // Verify the expected behavior of the above Act
+    EXPECT_EQ(expected_status, actual_status); // Example assertion
+    EXPECT_NEAR(expected_angle, actual_angle, CMD_RATIO); // 0.56 millidegree precision
+}
+TEST_F(ControllerTest, NominalCheck2) {
+    // Arrange
+    // Set up any necessary variables, mocks, etc
+    w_status_t expected_status = W_SUCCESS;
+    float expected_angle = -0.0028;
+    flight_phase_get_state_fake.return_val = STATE_ACT_ALLOWED;
+
+    float p_dyn = 13420.0f;
+    float coeff = 1.0f;
+    controller_gain_t controller_gain = {0};
+    float roll_state_arr[FEEDBACK_GAIN_NUM] = {0, 0, 0.001};
+
+    // Act
+    // Call the function to be tested
+
+    w_status_t actual_status = interpolate_gain(p_dyn, coeff, &controller_gain);
+
+    float actual_angle;
+    get_commanded_angle(controller_gain, roll_state_arr, &actual_angle);
+
+    // Assert
+    // Verify the expected behavior of the above Act
+    EXPECT_EQ(expected_status, actual_status); // Example assertion
+    EXPECT_NEAR(expected_angle, actual_angle, CMD_RATIO); // 0.56 millidegree precision
+}
+TEST_F(ControllerTest, NominalCheck3) {
+    // Arrange
+    // Set up any necessary variables, mocks, etc
+    w_status_t expected_status = W_SUCCESS;
+    float expected_angle = -0.0284;
+    flight_phase_get_state_fake.return_val = STATE_ACT_ALLOWED;
+
+    float p_dyn = 13420.0f;
+    float coeff = 1.0f;
+    controller_gain_t controller_gain = {0};
+    float roll_state_arr[FEEDBACK_GAIN_NUM] = {0.001, 0, 0.01};
+
+    // Act
+    // Call the function to be tested
+
+    w_status_t actual_status = interpolate_gain(p_dyn, coeff, &controller_gain);
+
+    float actual_angle;
+    get_commanded_angle(controller_gain, roll_state_arr, &actual_angle);
+
+    // Assert
+    // Verify the expected behavior of the above Act
+    EXPECT_EQ(expected_status, actual_status); // Example assertion
+
+    EXPECT_NEAR(expected_angle, actual_angle, CMD_RATIO); // 0.56 millidegree precision
+}
+
+TEST_F(ControllerTest, InterpolationOutOfBoundCheck) {
+    // Arrange
+    // Set up any necessary variables, mocks, etc
+    w_status_t expected_status = W_FAILURE;
+    float expected_angle = 0.0f;
+    flight_phase_get_state_fake.return_val = STATE_ACT_ALLOWED;
+
+    float p_dyn = 10.0f;
+    float coeff = 1.0f;
```

Comment:
I feel like you should change up `coeff` values for proper coverage. iirc it can be negative?